File: src/main/java/com/github/jamesgallicchio/stravum/JsonRPCSocket.java
package com.github.jamesgallicchio.stravum;

import com.thetransactioncompany.jsonrpc2.*;

import java.io.*;
import java.net.InetAddress;
import java.net.Socket;
import java.util.List;
import java.util.Map;
import java.util.function.Consumer;

public class JsonRPCSocket implements Closeable {

    private Socket s;
    private BufferedReader in;
    private BufferedWriter out;

    private List<Consumer<JSONRPC2Notification>> handlers;
    private Map<Object, JSONRPC2Response> waiting;

    private boolean polling;

    public JsonRPCSocket(String url, int port) {
        try {
            s = new Socket(InetAddress.getByName(url), port, InetAddress.getLocalHost(), 0);
            in = new BufferedReader(new InputStreamReader(s.getInputStream()));
            out = new BufferedWriter(new OutputStreamWriter(s.getOutputStream()));
        } catch (IOException e) {
            e.printStackTrace();
        }

        polling = true;
        new Thread(() -> {
            try {
                while (polling) {
                    try {
                        String incoming = in.readLine();
                        JSONRPC2Message m = JSONRPC2Message.parse(incoming);

                        if (m instanceof JSONRPC2Response) {
                            waiting.put(((JSONRPC2Response) m).getID(), (JSONRPC2Response) m);
                        } else if (m instanceof JSONRPC2Notification) {
                            for (Consumer<JSONRPC2Notification> handler : handlers) {
                                handler.accept((JSONRPC2Notification) m);
                            }
                        }
                    } catch (JSONRPC2ParseException ignored) {}
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
        }).run();
    }

    public JSONRPC2Response request(JSONRPC2Request r) throws IOException {
        out.write(r.toJSONString());
        out.flush();

        Object key = r.getID();
        try {
            while (!waiting.containsKey(key)) Thread.sleep(25);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        return waiting.remove(key);
    }

    public void addNotificationHandler(Consumer<JSONRPC2Notification> handler) {
        handlers.add(handler);
    }

    public void notify(JSONRPC2Notification n) throws IOException {
        out.write(n.toJSONString());
        out.flush();
    }

    public void close() {
        try {
            polling = false;
            in.close();
            out.close();
            s.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/scala/com/github/jamesgallicchio/stravum/jsonrpc/JsonRPCSocket.scala
package com.github.jamesgallicchio.stravum.jsonrpc

import java.io.{BufferedReader, BufferedWriter, InputStreamReader, OutputStreamWriter}
import java.net.{InetAddress, Socket}
import java.util.concurrent.{Executors, TimeUnit}

import com.github.jamesgallicchio.stravum.jsonrpc.Message.{Notify, Request, Response}
import monix.execution.Scheduler.Implicits.global
import monix.reactive.Observable
import play.api.libs.json.JsValue

import scala.collection.mutable
import scala.concurrent.duration.TimeUnit
import scala.concurrent.{Future, Promise}

class JsonRPCSocket(host: String, port: Int, notifyCacheMax: Int = 5) {
  private val socket = new Socket(InetAddress.getByName(host), port, InetAddress.getLocalHost, 0)
  private val outW = new BufferedWriter(new OutputStreamWriter(socket.getOutputStream))
  private object out {
    def write(s: String): Unit = {println(s"out: $s"); outW.write(s)}
    def flush(): Unit = outW.flush()
  }

  // Holds pending requests
  private val requests = mutable.Map[JsValue, Promise[Response]]()

  // Observes socket lines
  private val obs = Observable.fromLinesReader(new BufferedReader(new InputStreamReader(socket.getInputStream)))
    .dump("in: ").map(Message.parse).filter(_.isDefined).map(_.get).publish

  println("pre response handler")

  obs.collect { case r: Response => r }.foreach(r => requests.remove(r.id).foreach(_.success(r)))

  println("past response handler")

  // Process notifications
  val notifyStream: Observable[Notify] = obs.collect { case n: Notify => n }.cache(notifyCacheMax)

  println("past notify stream")

  private val executor = Executors.newScheduledThreadPool(10)

  def request(request: Request, timeout: Long = 15, unit: TimeUnit = TimeUnit.SECONDS): Future[Message.Response] = {
    // Store a promise to be responded to
    val promise = Promise[Response]
    requests.put(request.id, promise)

    // Send request
    out.write(request.toString)
    out.flush()

    // Remove request from list on timeout
    executor.schedule((() => requests.remove(request.id).foreach(_.failure(new InterruptedException()))): Runnable, timeout, unit)

    promise.future
  }

  def notify(notif: Notify): Unit = {
    out.write(notif.toString)
    out.flush()
  }
}